index.bs

```diff
@@ -1088,34 +1104,111 @@ strong map from handle ids to their corresponding objects.
 OwnershipModel = "root" / "none";
 </pre>
 
-<dfn>RemoteReference</dfn> represents a remote reference to an existing ECMAScript
-object in [=handle object map=] in the given [=Realm=].
+<dfn>RemoteReference</dfn> is either a <code>RemoteObjectReference</code>
+representing a remote reference to an existing ECMAScript object in
+[=handle object map=] in the given [=Realm=], or is a <code>SharedReference</code>
+representing a reference to a [=node=] in the given [=Realm=].
 
 <pre class="cddl remote-cddl local-cddl">
 Handle = text;
+SharedId = text;
+</pre>
 
+<pre class="cddl remote-cddl">
 RemoteReference = {
+  RemoteObjectReference //
+  SharedReference
+}
+
+RemoteObjectReference = {
    handle: Handle,
+   ?sharedId: SharedId
+   Extensible
+}
+
+SharedReference = {
+   ?handle: Handle,
+   sharedId: SharedId
    Extensible
 }
 </pre>
 
 Issue: handle "stale object reference" case.
 
+Note: if the provided reference has both <code>handle</code> and
+<code>sharedId</code>, the algorithm will first try to resolve <code>handle</code>,
+then <code>sharedId</code>, and if non of that succeeded, return a error.
+
+### Deserialization ### {#data-types-reference-deserialization}
+
 <div algorithm>
 To <dfn>deserialize remote reference</dfn> given |realm| and |remote reference|:
 
-1. Let |handle id| be the value of the <code>handle</code> field of |remote reference|.
+1. Let |remote value| be <code>null</code>.
+
+1. If |remote reference| matches the <code>RemoteObjectReference</code>
+   production, set |remote value| to the result of
+   [=deserialize remote object reference=] given |realm| and |remote reference|.
+
+1. If |remote value| is <code>null</code> and |remote reference| matches the
+  <code>SharedReference</code> production, set |remote value| to the result of
+  [=trying=] to [=deserialize shared reference=] given |realm| and
+  |remote reference|.
+
+1. If |remote value| is <code>null</code>, return [=error=] with [=error code=]
+   [=invalid argument=].
+
+1. Return [=success=] with data |remote value|.
+
+</div>
+
+<div algorithm>
+To <dfn>deserialize remote object reference</dfn> given |realm| and
+|remote object reference|:
+
+1. Let |handle id| be the value of the <code>handle</code> field of
+   |remote object reference|.
 
 1. Let |handle map| be |realm|'s [=handle object map=]
 
-1. If |handle map| does not contain |handle id|, then return [=error=] with
-   [=error code=] [=invalid argument=].
+1. If |handle map| does not contain |handle id|, then return <code>null</code>.
 
 1. Return |handle map|[|handle id|].
 
 </div>
 
+<div algorithm>
+To <dfn>deserialize shared reference</dfn> given |realm| and |shared reference|:
+
+1. Let |browsing context| be [=get the browsing context=] of the |realm|.
```

Comment:
Can we use "with" rather than "of the"? I think that it's nicer to generally try to call algorithms in a consistent way.